### ISSUE TITLE
kindaVIM crashes · KeyCode: 10/0xa

### ISSUE THREAD

**Issue by confluencepoint** [godbout/kindaVim.docs#112]
macOS: v12.5 (21G72)
kindaVim: v30

kindaVIM crashes when entering `KeyCode: 10/0xa`

![German Keyboard Layout](https://user-images.githubusercontent.com/890362/183280716-e4fb2069-ee54-4e81-b0f9-bb78884e9faa.png)

**Comment by godbout** [comment]
😲️

so it is when you're trying to get a `^` or the weird `o`? also hmm surprised that it would crash on a particular keycode. would expect more of a particular (bad) app. is it crashing on every app? if not, which one is the culprit? thanks

(sorry for the late answer.)

**Comment by confluencepoint** [comment]
The crash happens in any app.I press Escape and then a `^`. kV crashes immediately.

**Comment by godbout** [comment]
that's gonna be tough one. no issue here, haven't heard anything from other German keyboard users. i'll set my keyboard layout to Germän but i'd be surprised if it crashed.

is it new to kV30? you've been using kV for a while no? has this already crashed?

**Comment by confluencepoint** [comment]
I had previously installed kindaVim version 28+29. And I must confess to you that only this weekend I got to look at kV properly and test it extensively.

Meanwhile, I have been able to reproduce the bug with all apps installed on my Mac.

Even with `g^` kV crashes. 


_Can it possibly be related to the fact that the `^` character is marked so strangely (yellow)? And only after another keystroke loses the marking?_

![22080701](https://user-images.githubusercontent.com/890362/183309625-9ecf838d-4ca3-44b2-acc3-5e29b7a40b49.gif)

**Comment by godbout** [comment]
> I had previously installed kindaVim version 28+29. And I must confess to you that only this weekend I got to look at kV properly and test it extensively.

![IMG_2480](https://user-images.githubusercontent.com/121373/183421133-36546fdf-dc51-4d2f-b516-a7cf5a7359aa.jpeg)

> Meanwhile, I have been able to reproduce the bug with all apps installed on my Mac.
> 
> Even with `g^` kV crashes.
> 
> _Can it possibly be related to the fact that the `^` character is marked so strangely (yellow)? And only after another keystroke loses the marking?_
> 
> ![22080701](https://user-images.githubusercontent.com/890362/183309625-9ecf838d-4ca3-44b2-acc3-5e29b7a40b49.gif)

oh yes. most probably. what is this yellow thing? is this a normal thing happening when you type the `^` on a German keyboard? is there any way i can reproduce this on a US keyboard? (i'll check but asking coz i'm lazy.)

there's still places in kV's code that _expect_ the Accessibility API from Apple to return proper data. but i've got surprises with Electron app returning negative caret locations, etc. so i'm patching. the one you're having, if it crashes in every app, it has to be something related to how macOS handles the length of that strange character/yellow thing. i have to reproduce, but it should be ok to handle.

fun fact: dealing with emojis and kV. one emoji can be 2 characters, or 3, or 5, or 7, or 12, depending on the encoding, the emoji itself, etc. so when you want to do a `l` move for example, you can't just move 1 character. you need to calculate every move. that `^` thing feels a bit like this. lovely!

**Comment by godbout** [comment]
ok, had a try. so i have two questions:

1. how to you get your German keyboard layout on macOS. from all the German keyboards i have, none has the `^` on the left:

https://user-images.githubusercontent.com/121373/183469511-d23558a6-7d56-4425-b1c0-ae9da476b1b8.mp4

2. i've tried with the Swiss (sorry ☹️) German one, that has the `^` on the right and luckily does the yellow thingy. but it works here:

https://user-images.githubusercontent.com/121373/183469627-e9ee9778-6a7f-4376-a62e-83f05766b39e.mp4

so the second question is: you said kV "crashes". when? you don't even enter Normal Mode and it crashes? it crashes as soon as you enter Normal Mode? how do you enter Normal Mode? (escape, jk, etc.) what about if you enter Normal Mode not close to the `^`?

i need more precisions, coz it works here.

**Comment by godbout** [comment]
can you also attach/paste the crash report from the `Console.app`? the stack trace should be there. thanks.

**Comment by godbout** [comment]
ok i'm able to make kV crash sometimes with the French keyboard and that `^` where you need to add a space after (dead key). can't reproduce reliably yet, but it's a start.

**Comment by godbout** [comment]
ok so i've found the issue. so no more crash. but still, an important question: how to you use the `^` in Vim? or what do you expect? are you supposed to type `^` and then `space` in order to do the Vim `^` move?

**Comment by godbout** [comment]
so, crash should be gone in kV31. but i'll need your input about how do/will you use Vim moves with your German keyboard. as far as i understand now, lots of move will not work. not sure why though, as issues related to keyboard layouts have been solved in the past for quite some cases, so it should work for all (https://github.com/godbout/kindaVim.docs/issues/12, https://github.com/godbout/kindaVim.docs/issues/46). 

it's possible that i need to implement some more keyboard layout software mapping. see questions in comments above. thanks!

**Comment by godbout** [comment]
a14 is out and at least it shouldn't crash anymore.

**Comment by godbout** [comment]
closing as the crash is gone. if the move itself doesn't work the way you want/need, let's open another issue. thanks.
